### DISCUSSION TITLE
Laravel Sail share via expose doesn't seem to function at all

### DISCUSSION THREAD

**Discussion by Jamesking56** [laravel/framework#35987] [Q&A] [answered]
Hi guys,

Trying to share my Laravel site with my phone to do some testing via Laravel Sail.

So I type in `sail share` and get the usual output. When I visit the dashboard URL that appears on my Mac, I get Site cannot be reached.

When I try to visit the URL generated by sail locally and via my phone, I get `ERR_EMPTY_RESPONSE` in Chrome.

Not sure whats going on? I'm using Docker Desktop for Mac. Am I doing something wrong here?

**Comment by jearsonbgomez** [comment]
Hello,

Also in wsl2 not working. Even the the dashboard-url I am getting the ERR_CONNECTION_REFUSED.

**Comment by Jamesking56** [comment]
I've upgraded this to an actual bug report so hopefully it gets some visibility: https://github.com/laravel/framework/issues/36425

**Comment by Jamesking56** [comment, marked as answer]
A patch has been issued for Laravel Sail ready for the next release.

**Reply by jearsonbgomez** [reply]
@Jamesking56 Thanks.

**Comment by shortcodes** [comment]
Sorry guys but i am still failing to use it. Go newest `laravel/sail` version `v1.4.7`. Working on `wsl2` ubuntu18 distro and when i click on link provided by command - page is loading forever. In command line I can see my attempts to connect but loading goes forever and never stops. Do I do something wrong?

Regards!

**Reply by Jamesking56** [reply]
It looks like it has broken again for me too, I think 1.4.7 has undone the fix in 1.4.6.

**Comment by Lubje** [comment]
Couldn't get it work either with these versions:
- laravel/framework: v8.33.1
- laravel/sail: v1.4.8

**Reply by Jamesking56** [reply]
Can you comment on the bug issue please https://github.com/laravel/framework/issues/36425

**Comment by bjuppa** [comment]
Same here, I get sharing to work on Sail `v1.4.6` only, so I downgraded to that for now.

**Comment by bjuppa** [comment]
https://github.com/laravel/framework/issues/36425 has been updated with a description how to fix this!

Basically, make sure you clear out all your `beyondcodegmbh/expose-server` images from Docker before running `sail share` again. That will make sure a fresh image is downloaded with correct version.

**Reply by Lubje** [reply]
I'm using breeze and when I go to the generated url on my phone the first page works, but all the links (like the one for login) are pointed at 'host.docker.internal' instead of the generated one.

Do you also have this issue?

**Reply by bjuppa** [reply]
Actually, yes @Lubje... I also have that issue with the internal docker urls coming out of Laravel url-helpers, but I haven't yet researched it further as I've just needed `sail share` to give external services access to my local webhook endpoints during manual testing. So they do single isolated requests only, no browsing.

 If I was ever to use `sail share` to have another person browse remotely, I it wouldn't work for me either.

Maybe it's worth raising as a separate issue at https://github.com/laravel/sail/issues

**Reply by Lubje** [reply]
Thanks for the response.

Was thinking of creating an issue, but I want to investigate some more this weekend, before I trouble the team with it :)

**Reply by davidgittins** [reply]
I'm encountering the same issue. 

I've tried downgrading sail from 1.4.8 => 1.4.6 to no avail. Also removed beyondcodegmbh/expose-server images and re-tried - still the same problem.

**Reply by bjuppa** [reply]
@davidgittins could you see it downloading a new expose image when you ran `sail share`?

**Reply by davidgittins** [reply]
@bjuppa When I first ran `sail share` it automatically downloaded `beyondcodegmbh/expose-server:latest`. After I deleted this image and downgraded sail, it then downloaded `beyondcodegmbh/expose-server:1.5.1`.

It's worth mentioning that today was my first time using sail so I wouldn't have had any older historic images on my machine.

**Reply by bjuppa** [reply]
That makes me wonder if your problem may be different @davidgittins ... When you visit the url given by `sail share`, do you see anything in the log, in-between the countdowns? When I had the issue I could see the request coming in there. It'd look like this:


```
...
Remaining time: 00:29:53
Remaining time: 00:29:52
Remaining time: 00:29:51
+--------+-----+----------+---------------------+----------+
| Method | URI | Response | Time                | Duration |
+--------+-----+----------+---------------------+----------+
| GET    | /   |          | 2021-03-19 13:22:39 | 57ms     |
+--------+-----+----------+---------------------+----------+
Remaining time: 00:29:50
...
```

Do you see that at all? If not, I think you may have another issue.

**Reply by davidgittins** [reply]
The issue I have seems similar to what yourself and @Lubje have described where URLs/redirects are the special docker DNS name (host.docker.internal).

Here's what is output in my countdown if I try to visit the exposed domain:

```
+--------+------------+-----------+---------------------+----------+
| Method | URI        | Response  | Time                | Duration |
+--------+------------+-----------+---------------------+----------+
| GET    | /dashboard | 302 Found | 2021-03-19 13:34:37 | 64ms     |
| GET    | /          | 302 Found | 2021-03-19 13:34:36 | 90ms     |
+--------+------------+-----------+---------------------+----------+
Remaining time: 00:29:54
Remaining time: 00:29:53
```
We're being redirected from `/` to `/dashboard` as per a simple route but all that shows in the browser is http://host.docker.internal/login.

**Reply by davidgittins** [reply]
I can get a un-styled page by going directly to `/login` but any attempt to login will be met with that Docker URL.

Here an example <link> to the stylesheet that's failing:
`<link rel="stylesheet" href="http://host.docker.internal/css/app.css">`

**Reply by bjuppa** [reply]
Ah, sorry I misunderstood @davidgittins, yeah you're also seeing that issue with internal urls. First I thought you couldn't get access through the share url at all.

This problem with internal urls is present in every version of `laravel/sail` that I've tried, so it won't help downgrading, and it won't help deleting expose images from docker either 😞 

Let's see what @Lubje comes up with, then we might add an official issue!

**Reply by Lubje** [reply]
I just checked that I don't have any custom host additions in my project: no proxy, no custom APP_URL setting etc...
Als deleted all the images and rebuild them with the same result, so it seems we have a valid issue here?

**Reply by hstandeffer** [reply]
I'm also having the issue where expose points to the internal local URL instead of the remote. Would anyone want to add it to the main issues page for Sail? I used Laravel Breeze as well to generate the views if that helps at all

**Reply by adgower** [reply]
I am also having the same issue: "when I go to the generated url on my phone the first page works, but all the links (like the one for login) are pointed at 'host.docker.internal' instead of the generated one"

**Reply by GeorgII-web** [reply]
Same, Laravel's functions like route(), Redirect etc. generates 'http://host.docker.internal' instead of '*projectname*.laravel-sail.site:8080'.
Looks like docker container gives that domain to Laravel.

**Comment by GeorgII-web** [comment]
Hi, who has bug with 'host.docker.internal' in a request headers, it could helps.
Had created issue in Sail repo https://github.com/laravel/sail/issues/113

So, Laravel docs was patched by Laravel team:
```
When sharing your site via the `share` command, you should configure your application's trusted proxies within the `TrustProxies` middleware. Otherwise, URL generation helpers such as `url` and `route` will be unable to determine the correct HTTP host that should be used during URL generation:

    /**
     * The trusted proxies for this application.
     *
     * @var array|string|null
     */
    protected $proxies = '*';
```

**Reply by bjuppa** [reply]
Thank you very much @GeorgII-web!

**Reply by Lubje** [reply]
Thanks @GeorgII-web for creating the issue.

**Reply by s-marjanovic-bm** [reply]
This saved my life! Thank you!

**Comment by diegoldev** [comment]
Hi all. Seems that the error comes back with new Expose v2. Anyone facing same issue?

**Comment by robertsheacole** [comment]
I'm using valet for proxying my urls to the Docker containers. `valet share` still works.

**Comment by m33ts4k0z** [comment]
Hi all.

I believe new issues were introduced this year regarding the sail share command on WSL2.

I have also opened an issue [here](https://github.com/laravel/sail/issues/446)
